sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/JavaOrcColumnarBatchReader.java

```diff
@@ -0,0 +1,482 @@
+              for (int index = 0; index < batchSize; index++) {
+                toColumn.putByteArray(index, data.vector[0], data.start[0], data.length[0]);
+              }
+            } else if (type instanceof DecimalType) {
+              DecimalType decimalType = (DecimalType)type;
+              putDecimalWritables(
+                toColumn,
+                batchSize,
+                decimalType.precision(),
+                decimalType.scale(),
+                ((DecimalColumnVector)fromColumn).vector[0]);
+            } else {
+              throw new UnsupportedOperationException("Unsupported Data Type: " + type);
+            }
+          }
+        } else if (fromColumn.noNulls) {
+          DataType type = field.dataType();
+          if (type instanceof BooleanType) {
+            long[] data = ((LongColumnVector)fromColumn).vector;
+            for (int index = 0; index < batchSize; index++) {
+              toColumn.putBoolean(index, data[index] == 1);
+            }
+          } else if (type instanceof ByteType) {
+            long[] data = ((LongColumnVector)fromColumn).vector;
+            for (int index = 0; index < batchSize; index++) {
+              toColumn.putByte(index, (byte)data[index]);
+            }
+          } else if (type instanceof ShortType) {
+            long[] data = ((LongColumnVector)fromColumn).vector;
+            for (int index = 0; index < batchSize; index++) {
+              toColumn.putShort(index, (short)data[index]);
+            }
+          } else if (type instanceof IntegerType || type instanceof DateType) {
+            long[] data = ((LongColumnVector)fromColumn).vector;
+            for (int index = 0; index < batchSize; index++) {
+              toColumn.putInt(index, (int)data[index]);
+            }
+          } else if (type instanceof LongType) {
+            toColumn.putLongs(0, batchSize, ((LongColumnVector)fromColumn).vector, 0);
+          } else if (type instanceof TimestampType) {
+            TimestampColumnVector data = ((TimestampColumnVector)fromColumn);
+            for (int index = 0; index < batchSize; index++) {
+              toColumn.putLong(index, fromTimestampColumnVector(data, index));
+            }
+          } else if (type instanceof FloatType) {
+            double[] data = ((DoubleColumnVector)fromColumn).vector;
+            for (int index = 0; index < batchSize; index++) {
+              toColumn.putFloat(index, (float)data[index]);
+            }
+          } else if (type instanceof DoubleType) {
+            toColumn.putDoubles(0, batchSize, ((DoubleColumnVector)fromColumn).vector, 0);
+          } else if (type instanceof StringType || type instanceof BinaryType) {
+            BytesColumnVector data = ((BytesColumnVector)fromColumn);
+            for (int index = 0; index < batchSize; index++) {
+              toColumn.putByteArray(index, data.vector[index], data.start[index], data.length[index]);
```

Comment:
not related, but why ORC design the API like this? If the data is `byte[][]`, why we would ever need `start` and `length`?

Reply:
@cloud-fan Heh i just wrote about the same confusion at the exact same time. I checked some of our old code (we have custom ORC writer) and the start and length vectors are actually used on the other side: the writer can use single contiguous block and mark boundaries with those, exactly the same way i thought we could use in the read side.